mreg_cli/utilities/api.py

```diff
@@ -336,14 +342,69 @@ def get_list_unique(
     :returns: A single dictionary, or None if no result was found and ok404 is True.
     """
     ret = get_list_generic(path, params, ok404, expect_one_result=True)
-
-    if not isinstance(ret, dict):
-        raise CliError(f"Expected a single result, got {type(ret)}.")
-
     if not ret:
         return None
 
-    return ret
+    try:
+        validator = TypeAdapter(JsonMapping)
+        return validator.validate_python(ret)
+    except ValueError as e:
+        raise ValidationError(f"Failed to validate response from {path}: {e}") from e
+
+
+class PaginatedResponse(BaseModel):
+    """Paginated response data from the API."""
+
+    count: int
+    next: str | None
+    previous: str | None
+    results: list[Json]
+
+    @field_validator("count", mode="before")
+    @classmethod
+    def _none_count_is_0(cls, v: Any) -> Any:
+        """Ensure `count` is never `None`."""
+        # Django count doesn't seem to be guaranteed to be an integer.
+        # Ensures here that None is treated as 0.
+        # https://github.com/django/django/blob/bcbc4b9b8a4a47c8e045b060a9860a5c038192de/django/core/paginator.py#L105-L111
+        return v or 0
+
+    @classmethod
+    def from_response(cls, response: Response) -> PaginatedResponse:
+        """Create a PaginatedResponse from a Response."""
+        return cls.model_validate_json(response.text)
+
+
+ListResponse = TypeAdapter(list[Json])
+"""JSON list (array) response adapter."""
+
+
+# TODO: Provide better validation error introspection
+def validate_list_response(response: Response) -> list[Json]:
+    """Parse and validate that a response contains a JSON array.
+
+    :param response: The response to validate.
+    :raises ValidationError: If the response does not contain a valid JSON array.
+    :returns: Parsed response data as a list of Python objects.
+    """
+    try:
+        return ListResponse.validate_json(response.text)
+    # NOTE: ValueError catches custom Pydantic errors too
+    except ValueError as e:
+        raise ValidationError(f"{response.url} did not return a valid JSON array") from e
+
+
+def validate_paginated_response(response: Response) -> PaginatedResponse:
+    """Validate and parse that a response contains paginated JSON data.
+
+    :param response: The response to validate.
+    :raises ValidationError: If the response does not contain valid paginated JSON.
+    :returns: Parsed response data as a PaginatedResponse object.
+    """
+    try:
+        return PaginatedResponse.from_response(response)
+    except ValueError as e:
+        raise ValidationError(f"{response.url} did not return valid paginated JSON") from e
```

Comment:
Not too happy about catching `ValueError` instead of `ValidationError` here, but it's necessary since the `Json` type raises a `PydanticCustomError` in order to suppress an insanely large traceback from checking every union type:

```
>>> j.validate_python(["foo",object()])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/pederhan/repos/mreg-cli/.venv/lib/python3.12/site-packages/pydantic/type_adapter.py", line 260, in validate_python
    return self.validator.validate_python(object, strict=strict, from_attributes=from_attributes, context=context)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
pydantic_core._pydantic_core.ValidationError: 11 validation errors for nullable[union[dict[str,...],json-or-python[json=list[...],python=chain[is-instance[Sequence],function-wrap[sequence_validator()]]],str,int,float,bool]]
`dict[str,...]`
  Input should be a valid dictionary [type=dict_type, input_value=['foo', <object object at 0x102522610>], input_type=list]
    For further information visit https://errors.pydantic.dev/2.7/v/dict_type
`json-or-python[json=list[...],python=chain[is-instance[Sequence],function-wrap[sequence_validator()]]]`.1.`dict[str,...]`
  Input should be a valid dictionary [type=dict_type, input_value=<object object at 0x102522610>, input_type=object]
    For further information visit https://errors.pydantic.dev/2.7/v/dict_type
`json-or-python[json=list[...],python=chain[is-instance[Sequence],function-wrap[sequence_validator()]]]`.1.`json-or-python[json=list[...],python=chain[is-instance[Sequence],function-wrap[sequence_validator()]]]`
  Input should be an instance of Sequence [type=is_instance_of, input_value=<object object at 0x102522610>, input_type=object]
    For further information visit https://errors.pydantic.dev/2.7/v/is_instance_of
`json-or-python[json=list[...],python=chain[is-instance[Sequence],function-wrap[sequence_validator()]]]`.1.str
  Input should be a valid string [type=string_type, input_value=<object object at 0x102522610>, input_type=object]
    For further information visit https://errors.pydantic.dev/2.7/v/string_type
`json-or-python[json=list[...],python=chain[is-instance[Sequence],function-wrap[sequence_validator()]]]`.1.int
  Input should be a valid integer [type=int_type, input_value=<object object at 0x102522610>, input_type=object]
    For further information visit https://errors.pydantic.dev/2.7/v/int_type
`json-or-python[json=list[...],python=chain[is-instance[Sequence],function-wrap[sequence_validator()]]]`.1.float
  Input should be a valid number [type=float_type, input_value=<object object at 0x102522610>, input_type=object]
    For further information visit https://errors.pydantic.dev/2.7/v/float_type
`json-or-python[json=list[...],python=chain[is-instance[Sequence],function-wrap[sequence_validator()]]]`.1.bool
  Input should be a valid boolean [type=bool_type, input_value=<object object at 0x102522610>, input_type=object]
    For further information visit https://errors.pydantic.dev/2.7/v/bool_type
str
  Input should be a valid string [type=string_type, input_value=['foo', <object object at 0x102522610>], input_type=list]
    For further information visit https://errors.pydantic.dev/2.7/v/string_type
int
  Input should be a valid integer [type=int_type, input_value=['foo', <object object at 0x102522610>], input_type=list]
    For further information visit https://errors.pydantic.dev/2.7/v/int_type
float
  Input should be a valid number [type=float_type, input_value=['foo', <object object at 0x102522610>], input_type=list]
    For further information visit https://errors.pydantic.dev/2.7/v/float_type
bool
  Input should be a valid boolean [type=bool_type, input_value=['foo', <object object at 0x102522610>], input_type=list]
    For further information visit https://errors.pydantic.dev/2.7/v/bool_type
```

We probably don't want this to leak to the user, so that's why we raise the custom exception (and have to catch it accordingly).

Reply:
Yeah, makes sense. It's not ideal. Should we implement have an (error) log that we can have the user send us for debugging?

Reply:
Yeah, if we could store it somewhere that would be the best, but seeing as we print the URL along with the error, we IDEALLY should be able to find the invalid resource ourselves.